feat: add "Exploring Hardened Shielded Pools" blog post (#34)
ethereum/iptf-web: src/posts/2026-06-11-next-iteration-shielded-pools.md

---

> title: "The Next Iteration for Private Payments"

Reviewer: Maybe a slightly more neutral title would make sense? Sounds a bit salesy, which can be ok but maybe not for R&D blog

Author: Good point! Fixed in PR and in 271c39b

> description: "Shielded pools work today. Scaling them to private payments at Visa scale is a different problem. A look at what stands in the way, and a proof-of-concept that takes on two of them."

Reviewer: > A look at what stands in the way, and a proof-of-concept that takes on two of them.

I find this a bit confusing, maybe can rewrite with some more technical key words so people know what it is about?

Author: Fixed in 271c39b

date: 2026-06-11
author: "Yanis"
image: /assets/images/2026-06-11-next-iteration-shielded-pools/hero.png

> published: false

Reviewer: `published: false` is currently ignored by the Astro pages: `getCollection("posts")` is used directly in the blog index, homepage latest-posts section, and post detail paths, with no filter for `published` or future `date`.

That means merging this before 2026-06-11 will still publish the post immediately despite this frontmatter flag. If the intended behavior is “review now, publish later,” either keep the PR unmerged until June 11 or add a real filter before relying on this field.

Author: Fixed in 2ae7c97
Astro doesn't emulate release as Jekyll, removed the tag and switched to release on merge.

---

> Shielded pools are how you make a private payment on a public blockchain. Two earlier posts on this blog built one, first for [private bonds](/building-private-bonds-on-ethereum/), then for [private payments](/building-private-transfers-on-ethereum/). The primitive works. What comes next is a run of bigger iterations meant to make it both more resilient and more scalable: hard for any single party to watch, censor, or quietly take control of, and fast enough to carry payments at Visa scale. The trick is getting there without giving up the four things that make shielding worth having in the first place: censorship resistance, openness, privacy, and security.

Reviewer 2: Shielded pools are one way to make private payments on a public blockchain, though, right?

Author: Fixed in 271c39b

A shielded pool has a few moving parts, and each is where that gets hard. Through a wallet, it writes and reads the Ethereum contract state. And it proves, on the user's own device, the zero-knowledge proof that makes each spend valid, which is also where the post-quantum question bites. We will walk the open problems, then go deep on the two our prototype takes on.

## The open problems

> 

Reviewer: Maybe enumerate this a bit more? Like numbered list or so

Author: Fixed in 271c39b

> Shielded pools already move billions in value. The open question is whether they can carry private payments at scale: fast, cheap, and decentralized, the way a real payment network has to be. A few problems stand in the way.

Reviewer 2: "Shielded pools already move billions in value", is there a reference to this? as per my research, the combined tvl of compliant pools on ethereum is $100M or so. do you mean volume? if so volume is a statistic that is not very relevant, it can be generated :)

Author: Fixed in 271c39b

> On-chain state never stops growing. Every spend writes a nullifier to Ethereum, a short value that marks a note as spent so it can never be spent twice, and it has to stay there permanently, because the whole point is to remember forever. At payment volumes, say 150 million transactions a day, that is around 5 gigabytes a day in nullifiers alone, with no end date. [Bowe and Miers](https://eprint.iacr.org/2025/2031) work through the arithmetic, and the conclusion is plain: storing that much data on Ethereum forever, and asking everyone who validates a spend to keep consulting it, does not hold up. Epoch-based nullifiers are the most promising way out, though the tooling for them does not exist off the shelf yet.

Reviewer 2: ```suggestion
On-chain state never stops growing. Every spend writes a nullifier to Ethereum, a short value that marks a note as spent so it can never be spent twice, and it has to stay there permanently, because the whole point is to remember it forever. At payment volumes, say 150 million transactions a day, that is around 5 gigabytes a day in nullifiers alone, with no end date. [Bowe and Miers](https://eprint.iacr.org/2025/2031) work through the arithmetic, and the conclusion is plain: storing that much data on Ethereum forever, and asking everyone who validates a spend to keep consulting it, does not hold up. Epoch-based nullifiers are the most promising way out, though the tooling for them does not exist off the shelf yet.
```

a graph here to show growth would be ideal too

Author: Fiexd in 271c39b

> Censorship resistance is a problem of its own, with two ways to go wrong: paying for a spend, and getting it included. If you pay gas from your own address, you have tied yourself to your private spend, so today a relayer pays and submits for you. That works, but it makes the relayer a chokepoint that can stall, refuse, or be compelled to, and even past the relayer, whoever builds the block can simply leave your transaction out. Two proposed changes to Ethereum address both ends. Frame transactions with keyed two-dimensional nonces let a user pay and submit directly while keeping the gas payer separate from the spender. The FOCIL inclusion-list mechanism forces blocks to include eligible transactions, so a private spend cannot be quietly censored. ([EIP-8141](https://eips.ethereum.org/EIPS/eip-8141), [EIP-8250](https://ethereum-magicians.org/t/eip-8250-keyed-nonces-for-frame-transactions/28437), [EIP-7805](https://eips.ethereum.org/EIPS/eip-7805).)

Reviewer 2: FOCIL already has inclusion list in its acronym, perhaps you want to drop 

```suggestion
Censorship resistance is a problem of its own, with two ways to go wrong: paying for a spend, and getting it included. If you pay gas from your own address, you have tied yourself to your private spend, so today a relayer pays and submits for you. That works, but it makes the relayer a chokepoint that can stall, refuse, or be compelled to, and even past the relayer, whoever builds the block can simply leave your transaction out. Two proposed changes to Ethereum address both ends. Frame transactions with keyed two-dimensional nonces let a user pay and submit directly while keeping the gas payer separate from the spender. The FOCIL mechanism forces blocks to include eligible transactions, so a private spend cannot be quietly censored. ([EIP-8141](https://eips.ethereum.org/EIPS/eip-8141), [EIP-8250](https://ethereum-magicians.org/t/eip-8250-keyed-nonces-for-frame-transactions/28437), [EIP-7805](https://eips.ethereum.org/EIPS/eip-7805).)
```

Author: Fixed in 271c39b

The wallet carries more than you would think, and users feel it. Open an app you have not touched in a few weeks and it has to pull every recent note and try to decrypt each one before it can show a balance, because nothing on the outside says which notes are yours. On a busy pool that is minutes of waiting. And each time the wallet reads from the chain to build a spend, it does so through some provider that sees your address and your requests.

> Then there is the proof itself. Many shielded pools prove spends with Groth16, whose proofs are tiny and cheap for Ethereum to verify, but whose security a large quantum computer would break, and with it the soundness of every spend. The fix is a post-quantum proof system, which tends to be heavier, and that weight lands on the user's device, the same one already doing all that scanning. Nobody knows the timeline. Money is supposed to outlast the question.

Reviewer 2: using human article for an object here

i think this paragraph needs rewording

```suggestion
Proofs exist for shielded pools that use Groth16, which are small and cheap to verify on Ethereum but vulnerable to large quantum computers, compromising the soundness of every spend. A post-quantum proof system is needed, but it’s heavier and burdens the user’s device.
```

Author: Fixed in 271c39b

No design solves all of these at once, and pretending otherwise is how you ship something elegant and unusable. Our prototype takes two: the on-chain state, and the private read.

## What our prototype takes on

We built an extension on top of the working pool from the last post, keeping its note format and flows intact, and pointed it at two of those problems: keeping the on-chain state bounded, and letting a wallet read what it needs without anyone watching. The full [specification and implementation](https://github.com/ethereum/iptf-pocs/tree/master/pocs/private-payment/shielded-pool-extension) are open.

The rest of this post follows a spend through both, starting with the read.

## Reading without being watched

> Before a wallet can spend a note, it has to fetch the note's place in the commitment tree, the running list of every note ever created, and show the note is really in there. The proof that follows is airtight. The fetch is not. To get that data, the wallet asks a provider, a company like Infura that runs a node so you do not have to, and the request leaks twice. From your IP address, the provider learns you use the pool at all, no identity needed. From what you ask for, it can narrow down which note you are about to spend. The best on-chain privacy in the world can be undone by how the wallet reads around it.

Reviewer 2: maybe we dont want to mention a node provider here?

Author: Fixed in 271c39b

What we want is a way to read one row of a database without the operator learning which row. That exists.

### PIR, in plain terms

> It is called Private Information Retrieval, PIR, and the cleanest picture is a librarian. You want one specific book, but you would rather the librarian not know which, because the record of what you read is sensitive on its own. PIR lets you hand over a request, encrypted so the librarian can run the search across the whole library and pass back the right book, without ever learning which one it was. Under the hood, the database is a table of rows, you encrypt the index you want, the server does math over the entire table with your encrypted index, and you get back an encrypted answer only you can open. The server cannot tell row 5 from row 5,000.

Reviewer: I find "the cleanest picture is a librarian." language a bit confusing

Author: Fixed in 271c39b

> The catch, and it matters in a minute, is that PIR answers exactly one question: give me row i, where you already know i.

Reviewer: > The catch, and it matters in a minute,

This too

Maybe can do a pass over language over all? A few things like this

Author: Fixed in 271c39b

### Fetching a note's path blindly

> That is exactly the question our wallet needs to ask. When you received a note, you learned its position in the commitment tree from the payment that created it, so you know the index. To read its authentication path privately, we lay the tree out as a flat array, following [tree-pir](https://github.com/brech1/tree-pir), and the wallet sends a small batch of PIR queries, one per sibling on the path to the root, about log(N) of them for a tree of N notes. The server returns them and learns nothing about which leaf they belong to.

Reviewer: > That is exactly the question our wallet needs to ask. 

Same, humanizer skill should catch these

### What it costs

> None of this is free. PIR makes the server work over the whole database on every query, and schemes differ in how much precomputation each side keeps. The specification targets a recent scheme, InsPIRe, whose appeal is that it needs no per-database hint on the client. Our prototype ships an older, simpler one, SimplePIR, which gives the same privacy but does keep a hint on the client, derived from the public database. That hint is not a secret, so losing your phone loses nothing sensitive: you recompute or re-download it, and it has to be refreshed as the database grows. For real figures, the [PSE writeup on scaling Semaphore with PIR](https://pse.dev/projects/scaling-semaphore-pir) and the tree-pir benchmarks are the places to look. PIR is practical enough to build with today, and getting cheaper.

Reviewer 2: what is "per-database hint"?

Author: Fixed in 271c39b

## The server you don't have to trust

> So who is this server, and why trust it? Mostly you do not have to. On-chain, the contract keeps only a Merkle root, a single short fingerprint of every note. The server rebuilds the full tree from Ethereum's public event log and serves it, but it cannot lie, because the wallet checks whatever it returns against that root, and it gets the root from a light client that verifies it against Ethereum's consensus instead of trusting the server.

Reviewer: This slightly overstates what the PoC ships. The README says the light-client check ships the storage-proof verifier behind the `RootVerifier` port, but not a fully wired Helios light client; the in-process e2e trusts an anvil `stateRoot`.

Maybe qualify this as the target architecture, e.g. “In the target design, the wallet gets the root from a light client...” or add a short caveat that the PoC implements the proof verifier and leaves the consensus light-client wiring as production integration work.

Author: Fixed 271c39b

```
// The wallet trusts a root from consensus, not from the server.
root    = lightClient.verifyStorageSlot(pool, ROOT_SLOT)  // eth_getProof, checked against a finalized header

> rebuilt = hashUp(nodes_from_server)                        // reassemble the path the server returned

Reviewer 2: is `hashUp` a technical term here? i haven't seen it used

Author: Fixed in 271c39b

Now recall the other read-side problem, the cold wallet that takes minutes to find your money. It takes minutes because the wallet pulls every recent note and tries to decrypt each one, since nothing on the outside marks which are yours. You could ask the server to filter, but the thing that picks out your notes, your key, is exactly what you cannot hand over.

These are the same problem wearing two hats. In both, a server holds a big set, you need the few elements that match a private criterion, and revealing the criterion is the whole risk. That is private selection, and it is strictly harder than PIR's known-index fetch. The two are not identical: finding a neighbor is an ordering question over public values, finding your notes is an ownership question over ciphertexts. Same family, different test. Which is why no single primitive drops cleanly into both, why the same toolbox keeps reappearing, and why our prototype punts on the neighbor problem. Every project building private payments at scale runs into this, and reaches into the same toolbox to deal with it.

> ## The primitives, and who's building with them

Reviewer 2: Kohaku should have a reference here;

Author: Fixed in 271c39b

That shared problem, private selection, comes with a small toolbox, and the teams building private payments at scale each reach for a different part of it.

> Private Information Retrieval (PIR) reads a known row without revealing which one. That is the part our prototype uses, for the commitment path. Oblivious Message Retrieval (OMR) hands you the messages addressed to you without the server learning which are yours, aimed straight at note discovery. Fuzzy Message Detection (FMD) is the lighter cousin: it flags your notes plus a tunable share of decoys, so you trial-decrypt a smaller, blurred set, and [Penumbra](https://protocol.penumbra.zone/main/crypto/fmd.html) runs it in production.

Reviewer 2: i think we can just use the acronym for PIR here, it was already expanded earlier

Author: Fixed in 271c39b
